api/migrations/2020-07-25-172603_add_permission_models/up.sql

```diff
@@ -0,0 +1,40 @@
+CREATE TYPE role_type AS ENUM ('admin', 'spec_creator');
+CREATE TABLE roles (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    name role_type NOT NULL UNIQUE,
+    is_admin BOOLEAN NOT NULL DEFAULT FALSE -- admins implicitly get all permissions
+);
+
+CREATE TABLE user_roles (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    user_id UUID NOT NULL REFERENCES users(id),
+    role_id UUID NOT NULL REFERENCES roles(id),
+    UNIQUE(user_id, role_id)
+);
+
+CREATE TYPE permission_type AS ENUM (
+    'create_specs',
+    'modify_all_specs',
+    'delete_all_specs'
+);
+CREATE TABLE permissions (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    name permission_type NOT NULL UNIQUE
+);
+
+CREATE TABLE role_permissions (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    role_id UUID NOT NULL REFERENCES roles(id),
+    permission_id UUID NOT NULL REFERENCES permissions(id),
+    UNIQUE(role_id, permission_id)
+);
```

Comment:
@JRMurr how do you feel about these being enums? It's nice cause it makes it easy to map to Rust enums and make sure we don't get invalid names used. But idk how migration will be if we want to add/remove options.

Reply:
I like the enum, looks like we could use https://www.postgresql.org/docs/9.5/sql-altertype.html when we migrate